cpp/src/arrow/array.h

```diff
@@ -35,13 +35,88 @@
 
 namespace arrow {
 
+using BufferVector = std::vector<std::shared_ptr<Buffer>>;
+
+// When slicing, we do not know the null count of the sliced range without
+// doing some computation. To avoid doing this eagerly, we set the null count
+// to -1 (any negative number will do). When Array::null_count is called the
+// first time, the null count will be computed. See ARROW-33
+constexpr int64_t kUnknownNullCount = -1;
+
 class MemoryPool;
-class MutableBuffer;
 class Status;
 
 template <typename T>
 struct Decimal;
 
+// ----------------------------------------------------------------------
+// Generic array data container
+
+/// \brief Container for generic array data
+struct ARROW_EXPORT ArrayData {
```

Comment:
Done (so more explanation for why this is useful is there). I also moved this to an `arrow::internal` namespace and made it clear that this is for internal use and developers of Arrow. Should the need arise to make this a public API we can revisit